cpp/src/arrow/array.h

```diff
@@ -138,28 +173,26 @@ class ARROW_EXPORT NullArray : public Array {
  public:
   using TypeClass = NullType;
 
-  NullArray(const std::shared_ptr<DataType>& type, int32_t length)
-      : Array(type, length, length, nullptr) {}
-
-  explicit NullArray(int32_t length) : NullArray(std::make_shared<NullType>(), length) {}
+  explicit NullArray(int32_t length);
 
   Status Accept(ArrayVisitor* visitor) const override;
-};
 
-Status ARROW_EXPORT GetEmptyBitmap(
-    MemoryPool* pool, int32_t length, std::shared_ptr<MutableBuffer>* result);
+  std::shared_ptr<Array> Slice(int32_t offset, int32_t length) const override;
+};
 
 /// Base class for fixed-size logical types
 class ARROW_EXPORT PrimitiveArray : public Array {
  public:
-  virtual ~PrimitiveArray() {}
```

Comment:
I actually spent a good bit of time googling about this. 

My understanding is that if the base class has a virtual destructor, then it is not necessary to provide trivial destructor implementations in subclasses. If you forget to implement a virtual dtor in the *base* class, then destructing a base class instance has undefined behavior. 

If this is not true, we should really find a concrete reference. The reason I looked into it was because I see a number of duplicate symbols in `libarrow.so` from running `nm -g`. I'm not really sure why that is, at some point we should figure that out